### PR TITLE
Averages to cell centers

### PR DESCRIPTION
**Pull request by EZoni**
**Note** PR related to issues #751 and #797.

**Goal**

Unify and replace the AMReX functions `average_node_to_cellcenter`, `average_edge_to_cellcenter` and `average_face_to_cellcenter` (defined in `Src/Base/AMReX_MultiFabUtil.cpp` in the AMReX repository and called multiple times within the function `AverageAndPackVectorField` in `Source/Diagnostics/FieldIO.cpp`) with one single function handling general staggering of the input fields.

**Description**

The new namespace `Average` is defined in the new file **Source/Diagnostics/Average.H**. The namespace contains two overloads of the function `ToCellCenter`, computing the cell-centered averages of the data contained in a given input `MultiFab` and storing them in a given output `MultiFab`. The two overloads are implemented in the new file **Source/Diagnostics/Average.cpp**. The function `AverageAndPackVectorField` in the file **Source/Diagnostics/FieldIO.cpp** has been significantly simplified, replacing all calls to the AMReX functions `average_node_to_cellcenter`, `average_edge_to_cellcenter` and `average_face_to_cellcenter` with calls to the new function `ToCellCenter` and removing unnecessary `if` control statements handling the different types of staggering. The auxiliary function `PackPlotDataPtrs` is not used anymore and has been deleted. Extensive Doxygen documentation of the new namespace has been added in the header file **Source/Diagnostics/Average.H**.

**Testing the PR**

Travis CI tests pass, regression tests on CPUs (Battra) and GPUs (Garuda) pass:
- https://travis-ci.org/github/ECP-WarpX/WarpX/builds/663715886
- https://ccse.lbl.gov/pub/RegressionTesting/WarpX/2020-03-17-001/index.html
- https://ccse.lbl.gov/pub/GpuRegressionTesting/WarpX/2020-03-17-002/index.html

### CODE REVIEW

File: Source/Diagnostics/FieldIO.cpp [hunk 1]
```diff
@@ -254,6 +253,163 @@ PackPlotDataPtrs (Vector<const MultiFab*>& pmf,
...
+                           Array4<Real const> const& mf_in_arr,
+                           const IntVect stag,
+                           const int dcomp,
+                           const int scomp=0,
+                           const int ncomp=0 )
+{
+    const auto lo = lbound(bx);
+    const auto hi = ubound(bx);
+    int n = 0;
+    do {
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]: Could be for loop.

File: Source/Diagnostics/FieldIO.cpp [hunk 2]
```diff
@@ -254,6 +253,163 @@ PackPlotDataPtrs (Vector<const MultiFab*>& pmf,
...
+ */
+AMREX_GPU_HOST_DEVICE
+AMREX_FORCE_INLINE
+void AverageToCellCenter ( Box const& bx,
+                           Array4<Real> const& mf_out_arr,
+                           Array4<Real const> const& mf_in_arr,
+                           const IntVect stag,
+                           const int dcomp,
+                           const int scomp=0,
+                           const int ncomp=0 )
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
                           const int ncomp=1 )
```

File: Source/Diagnostics/FieldIO.cpp [hunk 3]
```diff
@@ -254,6 +253,163 @@ PackPlotDataPtrs (Vector<const MultiFab*>& pmf,
...
+#endif
+    // Loop over boxes (or tiles if not on GPU)
+    for (MFIter mfi( mf_out, TilingIfNotGPU() ); mfi.isValid(); ++mfi)
+    {
+        const Box bx = mfi.growntilebox( ngrow );
+        Array4<Real> const& mf_out_arr = mf_out.array( mfi );
+        Array4<Real const> const& mf_in_arr = mf_in->const_array( mfi );
+        const IntVect stag = mf_in->boxArray().ixType().ixType();
+        AMREX_LAUNCH_HOST_DEVICE_LAMBDA ( bx, tbx,
+        { AverageToCellCenter( tbx, mf_out_arr, mf_in_arr, stag, dcomp, scomp, ncomp ); });
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]: `ParallelFor`

File: Source/Diagnostics/FieldIO.cpp [hunk 4]
```diff
@@ -254,6 +253,163 @@ PackPlotDataPtrs (Vector<const MultiFab*>& pmf,
...
+ */
+AMREX_GPU_HOST_DEVICE
+AMREX_FORCE_INLINE
+Real AverageToCellCenter ( Array4<Real const> const& mf_in_arr,
+                           const IntVect stag,
+                           AMREX_D_DECL( int i, int j, int k ),
+                           const int comp=0 )
+{
+#if ( AMREX_SPACEDIM == 2 )
+    return 0.25 * ( mf_in_arr(i        ,j        ,0,comp)
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
    return 0.25_rt * ( mf_in_arr(i        ,j        ,0,comp)
```
everywhere.

File: Source/Diagnostics/FieldIO.cpp [hunk 5]
```diff
@@ -254,6 +253,163 @@ PackPlotDataPtrs (Vector<const MultiFab*>& pmf,
...
+ *                       , containing the data to be averaged
+ * \param[in]     ncomp  optional number of components to loop over for both the
+ *                       input and the output
+ *                       <a href="https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Array4.html">Array4</a>
+ *                       objects, extracted from the respective
+ *                       <a href="https://amrex-codes.github.io/amrex/doxygen/classamrex_1_1MultiFab.html">MultiFabs</a>
+ */
+void
+AverageToCellCenter ( MultiFab& mf_out,
+                      const MultiFab*& mf_in,
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
                      const MultiFab& mf_in,
```

File: Source/Diagnostics/FieldIO.cpp [hunk 6]
```diff
@@ -284,12 +442,12 @@ AverageAndPackVectorField( MultiFab& mf_avg,
...
     } else if ( vector_field[0]->is_nodal() ){
 
-        amrex::average_node_to_cellcenter( mf_avg, dcomp  ,
-                                          *vector_field[0], 0, 1, ngrow);
-        amrex::average_node_to_cellcenter( mf_avg, dcomp+1,
-                                          *vector_field[1], 0, 1, ngrow);
-        amrex::average_node_to_cellcenter( mf_avg, dcomp+2,
-                                          *vector_field[2], 0, 1, ngrow);
+        tmp_ptr = vector_field[0].get();
+        AverageToCellCenter( mf_avg, tmp_ptr, dcomp  , ngrow, 0, 1 );
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
        AverageToCellCenter( mf_avg, *vector_field[0], dcomp  , ngrow, 0, 1 );
```

File: Source/Diagnostics/FieldIO.cpp [hunk 7]
```diff
@@ -312,15 +470,19 @@ AverageAndPackVectorField( MultiFab& mf_avg,
             mf_total[2].reset(new MultiFab(vector_field[2]->boxArray(), dm, 1, vector_field[2]->nGrowVect()));
             ConstructTotalRZField(mf_total, vector_field);
             PackPlotDataPtrs(srcmf, mf_total);
-            amrex::average_face_to_cellcenter( mf_avg, dcomp, srcmf, ngrow);
+            AMREX_D_TERM( AverageToCellCenter( mf_avg, srcmf[0], dcomp  , ngrow );,
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
            AMREX_D_TERM( AverageToCellCenter( mf_avg, *(srcmf[0]), dcomp  , ngrow );,
```

File: Source/Diagnostics/FieldIO.cpp [hunk 8]
```diff
@@ -403,9 +569,8 @@ AverageAndPackScalarField( MultiFab& mf_avg,
         MultiFab::Copy( mf_avg, scalar_field, 0, dcomp, 1, ngrow);
         // - Fully nodal
     } else if ( scalar_field.is_nodal() ){
-
-        amrex::average_node_to_cellcenter( mf_avg, dcomp, scalar_field, 0, 1, ngrow);
-
+        const MultiFab* tmp_ptr = &scalar_field;
+        AverageToCellCenter( mf_avg, tmp_ptr, dcomp, ngrow, 0, 1 );
```
Comment (MaxThevenet) [inline Source/Diagnostics/FieldIO.cpp]:
```suggestion
        AverageToCellCenter( mf_avg, scalar_field, dcomp, ngrow, 0, 1 );
```

**Comment by dpgrote** [comment]
Thanks for this PR! Note that there will be many conflicts with PR#770. I would recommend merging #770 in first before continuing this PR. That PR has some simplifications of the AverageAndPackVectorField routine (in how the RZ case is treated). With those changes, this PR could even further simplify AverageAndPackVectorField.

Another comment - is there an advantage to having three layers of calls to AverageToCellCenter? I think it would be more succinct (and readable) to have it all in one routine. Also, I second the recommendation of using ParallelFor instead of explicit for loops over the box.

**Comment by dpgrote** [comment]
The routine AverageAndPackVectorField can be further simplified since now there are only two cases. If fully cell centered, then only a copy is needed. For all other cases, call the Average::ToCellCenter for each component. Also, the AMREX_D_TERM should not be used since in 2D there are still three components. And also the separate if block for 2D is no longer needed.

**Comment by EZoni** [comment]
> The routine AverageAndPackVectorField can be further simplified since now there are only two cases. If fully cell centered, then only a copy is needed. For all other cases, call the Average::ToCellCenter for each component. Also, the AMREX_D_TERM should not be used since in 2D there are still three components. And also the separate if block for 2D is no longer needed.

Thank you for the feedback @dpgrote , I'll fix these issues.

File: Source/Diagnostics/Average.H [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+     * \param[in] j         index along y to access the
+     *                      <a href="https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Array4.html">Array4</a>
+     *                      \c mf_in_arr containing the data to be averaged
+     * \param[in] k         index along z to access the
+     *                      <a href="https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Array4.html">Array4</a>
+     *                      \c mf_in_arr containing the data to be averaged
+     * \param[in] comp      index along the fourth component of the
+     *                      <a href="https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Array4.html">Array4</a>
+     *                      \c mf_in_arr containing the data to be averaged
+     */
```
Comment (MaxThevenet) [inline Source/Diagnostics/Average.H]:
```suggestion
     * \return averaged field at cell (I,j,k)
     */
```

File: Source/Diagnostics/Average.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+using namespace amrex;
+
+AMREX_GPU_HOST_DEVICE
+AMREX_FORCE_INLINE
+Real Average::ToCellCenter ( Array4<Real const> const& mf_in_arr,
+                             const IntVect stag,
+                             int i,
+                             int j,
+                             int k,
```
Comment (MaxThevenet) [inline Source/Diagnostics/Average.cpp]:
```suggestion
                             int const i,
                             int const j,
                             int const k,
```
these are passed by value, so adding `const` does not necessarily help with performance, but it clarifies the intent to other developers.

File: Source/Diagnostics/Average.cpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+                             const int comp)
+{
+    int sx = stag[0];
+    int sy = stag[1];
+#if   (AMREX_SPACEDIM == 2)
+    int sz = 0;
+#elif (AMREX_SPACEDIM == 3)
+    int sz = stag[2];
```
Comment (MaxThevenet) [inline Source/Diagnostics/Average.cpp]:
```suggestion
    int const sx = stag[0];
    int const sy = stag[1];
#if   (AMREX_SPACEDIM == 2)
    int constexpr sz = 0;
#elif (AMREX_SPACEDIM == 3)
    int const sz = stag[2];
```

**Pull request merged**